library/core/src/slice/sort/unstable/mod.rs

```diff
@@ -52,6 +58,51 @@ pub fn sort<T, F: FnMut(&T, &T) -> bool>(v: &mut [T], is_less: &mut F) {
     }
 }
 
+/// Unstable partial sort the range `start..end`, after which it's guaranteed that:
+///
+/// 1. Every element in `v[..start]` is smaller than or equal to
+/// 2. Every element in `v[start..end]`, which is sorted, and smaller than or equal to
+/// 3. Every element in `v[end..]`.
+#[inline(always)]
+pub fn partial_sort<T, F, R>(v: &mut [T], range: R, mut is_less: F)
+where
+    F: FnMut(&T, &T) -> bool,
+    R: RangeBounds<usize>,
+{
+    // Arrays of zero-sized types are always all-equal, and thus sorted.
+    if T::IS_ZST {
+        return;
+    }
+
+    let len = v.len();
+    let Range { start, end } = slice::range(range, ..len);
+
+    if end - start <= 1 {
+        // Empty range or single element. This case can be resolved in at most
+        // single partition_at_index call, without further sorting.
+
+        if start != len && end != 0 {
+            partition_at_index(v, start, &mut is_less);
+        } else {
+            // Do nothing if it is an empty range at start or end: all guarantees
+            // are already upheld.
+        }
+
+        return;
```

Comment:
Nit: I think the condition reads easier and this flows a bit better than the empty `else` branch for a comment.

```rust
if end == 0 || start == len {
    // Do nothing if it is an empty range at start or end: all guarantees
    // are already upheld.
    return;
}

partition_at_index(v, start, &mut is_less);
return;
```